src/Common/src/System/IO/PathInternal.cs

```diff
@@ -98,16 +98,100 @@ internal static StringBuilder TrimEnd(this StringBuilder builder, params char[]
         internal static int FindFileNameIndex(string path)
         {
             Debug.Assert(path != null);
-            PathInternal.CheckInvalidPathChars(path);
-            
+            CheckInvalidPathChars(path);
+
             for (int i = path.Length - 1; i >= 0; i--)
             {
                 char ch = path[i];
                 if (IsDirectoryOrVolumeSeparator(ch))
                     return i + 1;
             }
-            
+
             return 0; // the whole path is the filename
         }
+
+        /// <summary>
+        /// Returns true if the path ends in a directory separator.
+        /// </summary>
+        internal static bool EndsInDirectorySeparator(string path)
+        {
+            if (string.IsNullOrEmpty(path))
+                return false;
+
+            return IsDirectorySeparator(path[path.Length - 1]);
+        }
+
+        /// <summary>
+        /// Get the common path length from the start of the string.
+        /// </summary>
+        internal static int GetCommonPathLength(string first, string second, bool ignoreCase)
+        {
+            int commonChars = EqualStartingCharacterCount(first, second, ignoreCase: ignoreCase);
+
+            // If nothing matches
+            if (commonChars == 0)
+                return commonChars;
+
+            // Or we're a full string and equal length or match to a separator
+            if (commonChars == first.Length
+                && (commonChars == second.Length || IsDirectorySeparator(second[commonChars])))
+                return commonChars;
+
+            if (commonChars == second.Length && IsDirectorySeparator(first[commonChars]))
+                return commonChars;
+
+            // It's possible we matched somewhere in the middle of a segment e.g. C:\Foodie and C:\Foobar.
+            while (commonChars > 0 && !IsDirectorySeparator(first[commonChars - 1]))
+                commonChars--;
+
+            return commonChars;
+        }
+
+        /// <summary>
+        /// Gets the count of common characters from the left optionally ignoring case
+        /// </summary>
+        unsafe internal static int EqualStartingCharacterCount(string first, string second, bool ignoreCase)
```

Comment:
Is it possible to substitute this function with `string.Compare` or `string.CompareOrdinal` instead? The builtin string functions will typically be faster than hand-rolled pointer arithmetic. For example, [`CompareOrdinal`](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/String.Comparison.cs#L104-L110) compares 4 chars at a time and uses loop unrolling.